### PR TITLE
Add ssh config

### LINKED ISSUE #441: Add support for ietf-ssh-server YANG
The [recently ratified IETF RFC model for SSH](https://www.rfc-editor.org/rfc/rfc9644.html) provides an [server grouping](https://github.com/YangModels/yang/blob/dc5185dd28e15582434776a6a975123d13f638a2/standard/ietf/RFC/ietf-ssh-server%402024-10-10.yang#L149) that can be included in the Infix services model.

Scope:

- [x] Enable disable service
- [x] Change listening port
- [x] `ListenAddress`[^1]
- [x] Host key similar to NETCONF `"genkey"` setup

Future:

- [ ] Bind to a single interface
- [ ] Add support for certificates

> **Note:** [netopeer2-server use libssh](#440) for access on port 830.  This should probably be handled by the NETCONF model and not the SSH server model.

[^1]: which may also require `sysctl` setting `net.ipv4.ip_nonlocal_bind=1`, e.g., if binding to a dynamic address like 169.254.1.1, in very specific conditions -- one customer setup uses this for predictive addresses in a local setup with containers.

### PR DESCRIPTION
**Pull request by mattiaswal**
## Description
This fix #441 
But implemented in infix-services instead (without ietf-ssh)
## Checklist

Tick *relevant* boxes, this PR is-a or has-a:

- [ ] Bugfix
  - [ ] Regression tests
  - [ ] ChangeLog updates (for next release)
- [x] Feature
  - [x] YANG model change => revision updated?
  - [x] Regression tests added?
  - [x] ChangeLog updates (for next release)
  - [x] Documentation added?
- [x] Test changes
  - [ ] Checked in changed Readme.adoc (make test-spec)
  - [ ] Added new test to group Readme.adoc and yaml file
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (please detail in commit messages)
- [ ] Build related changes
- [ ] Documentation content changes
  - [ ] ChangeLog updated (for major changes)
- [ ] Other (please describe):

### CODE REVIEW

**Comment by mattiaswal** [comment]
<pre>

module: infix-services
  +--rw mdns
  |  +--rw enabled?   boolean
  +--rw ssh
  |  +--rw enabled?   boolean
  |  +--rw hostkey*   ks:asymmetric-key-ref
  |  +--rw listen* [name]
  |     +--rw name       string
  |     +--rw address?   inet:ip-address
  |     +--rw port?      inet:port-number
  +--rw web
     +--rw enabled?     boolean
     +--rw console
     |  +--rw enabled?   boolean
     +--rw netbrowse
     |  +--rw enabled?   boolean
     +--rw restconf
        +--rw enabled?   boolean
</pre>

File: src/confd/src/infix-services.c [hunk 1]
```diff
@@ -279,6 +292,72 @@ static int restconf_change(sr_session_ctx_t *session, uint32_t sub_id, const cha
...
+	}
+
+	LY_LIST_FOR(lydx_get_child(ssh, "listen"), listen) {
+		const char *address, *port;
+		int ipv6;
+		address = lydx_get_cattr(listen, "address");
+		ipv6 = !!strchr(address, ':');
+		port = lydx_get_cattr(listen, "port");
+		fprintf(fp, "ListenAddress %s%s%s:%s\n", ipv6 ? "[" : "", address, ipv6 ? "]" : "", port);
+	}
```
Comment (troglobit) [inline src/confd/src/infix-services.c:352]: Please add some whitespace (empty lines) between variable declarations, and invariants, for readability.  E.g., all if-checks for error handling should be surrounded by newlines.

File: src/confd/src/infix-services.c [hunk 2]
```diff
@@ -307,6 +386,128 @@ static int web_change(sr_session_ctx_t *session, uint32_t sub_id, const char *mo
...
+ */
+
+int rm(const char *path) {
+    struct dirent *entry;
+    struct stat statbuf;
+    DIR *dir;
+    if (!path) {
+	    errno = EINVAL;
+	    return 1;
+    }
```
Comment (troglobit) [inline src/confd/src/infix-services.c:403]: Same here, add missing whitespace and make sure to have start brace on a new line for functions :-)

File: src/confd/src/infix-services.c [hunk 3]
```diff
@@ -307,6 +386,128 @@ static int web_change(sr_session_ctx_t *session, uint32_t sub_id, const char *mo
...
+
+    if (!S_ISDIR(statbuf.st_mode))
+	    return remove(path);
+
+    dir = opendir(path);
+    if (!dir) {
+	    return 1;
+    }
+
```
Comment (troglobit) [inline src/confd/src/infix-services.c:416]: Minor, nitpick -- select one style: either always use braces for all if/for/when with single statement, or always skip braces for single statements.

File: src/confd/src/infix-services.c [hunk 4]
```diff
@@ -307,6 +386,128 @@ static int web_change(sr_session_ctx_t *session, uint32_t sub_id, const char *mo
...
+    return rmdir(path);
+}
+
+/* Store SSH public/private keys */
+static int change_keystore_cb(sr_session_ctx_t *session, uint32_t sub_id, const char *module_name,
+		     const char *xpath, sr_event_t event, uint32_t request_id, void *_)
+{
+	int rc = SR_ERR_OK;
+	sr_data_t *cfg;
+	struct lyd_node *changes, *change;
```
Comment (troglobit) [inline src/confd/src/infix-services.c:453]: same here, missing whitespace between variable declarations and code.

File: src/confd/yang/infix-services.yang [hunk 5]
```diff
@@ -3,10 +3,33 @@ module infix-services {
...
 
+  import ietf-inet-types {
+    prefix inet;
+    reference
+      "RFC 6991: Common YANG Data Types";
+  }
+  import ietf-crypto-types {
+    prefix ct;
+    reference
+      "RFC AAAA: YANG Data Types and Groupings for Cryptography";
```
Comment (troglobit) [inline src/confd/yang/infix-services.yang:14]: RFC AAAA?

File: src/confd/yang/infix-services.yang [hunk 6]
```diff
@@ -36,6 +59,49 @@ module infix-services {
       type boolean;
     }
   }
+  container ssh {
+    description "Configuration for the SSH daemon";
+    leaf enabled {
+      description "Disable or enable SSH daemon";
+      type boolean;
+      must ". = 'false' or (count(../listen) > 0 and count(../hostkey) > 0)" {
+        error-message "When SSH is enabled, there must be at least one listen address configured and a private key must be set.";
```
Comment (troglobit) [inline src/confd/yang/infix-services.yang:68]: Simplify: "Must be at least one listen address and a private key."

File: src/confd/yang/infix-services.yang [hunk 7]
```diff
@@ -36,6 +59,49 @@ module infix-services {
...
+      }
+    }
+
+    leaf-list hostkey {
+      description
+        "A reference to an asymmetric key that exists in
+          the central keystore.";
```
Comment (troglobit) [inline src/confd/yang/infix-services.yang:75]: For the sanity of people reading help texts.  The first sentence should be complete and without line breaks and kept within 70 chars.

Suggest simplify:

    description "Reference to asymmetric key in central keystore.

                 The hostkey can be shared with NETCONF, by default 'genkey' is used."

File: src/confd/yang/infix-services.yang [hunk 8]
```diff
@@ -36,6 +59,49 @@ module infix-services {
...
+    list listen {
+      key name;
+      leaf name {
+        type string;
+      }
+
+      leaf address {
+        description
+          "The local IP address to listen on for incoming
+           SSH client connections.  INADDR_ANY (0.0.0.0) or
```
Comment (troglobit) [inline src/confd/yang/infix-services.yang:93]: Same here, make it easier for users of the CLI help system.  Suggestion to simplify first sentence and then break up the rest in a separate paragraph with emtpy line between. (See above)

    description "Local IP address to listen on for inbound SSH connections.

                 INADDR_ANY ... "

File: src/confd/yang/infix-services.yang [hunk 9]
```diff
@@ -36,6 +59,49 @@ module infix-services {
...
+           INADDR6_ANY (0:0:0:0:0:0:0:0 a.k.a. ::) MUST be
+           used when the server is to listen on all IPv4 or
+           IPv6 addresses, respectively.";
+        type inet:ip-address;
+      }
+      leaf port {
+        description "Local port for SSH daemon to listen to.";
```
Comment (troglobit) [inline src/confd/yang/infix-services.yang:100]: To listen _on_, or to listen _to_?   We should use the same preposition for this and the `address` description.

File: test/infamy/netconf.py [hunk 10]
```diff
@@ -397,8 +397,8 @@ def delete_xpath(self, xpath):
         libyang.xpath_del(new, xpath)
 
         mod = self.ly.get_module(module)
-        oldd = mod.parse_data_dict(old, no_state=True)
-        newd = mod.parse_data_dict(new, no_state=True)
+        oldd = mod.parse_data_dict(old, no_state=True, validate=False)
+        newd = mod.parse_data_dict(new, no_state=True, validate=False)
```
Comment (troglobit) [inline test/infamy/netconf.py:401]: Heh, I think I have the same fix in my branch! :rofl: I have an additional check though before this, but we can take that on my PR later on.

Nice catch!

**Review by troglobit — CHANGES_REQUESTED** [review]
Amazing work on this, it looks really good!  So happy to get this support into Infix :heart:

File: src/confd/src/infix-services.c [hunk 11]
```diff
@@ -370,6 +451,129 @@ static int web_change(sr_session_ctx_t *session, uint32_t sub_id, const char *mo
...
 }
 
+/**
+ * Recursive directory remove (rm -r)
+ * @param file or directory to remove
+ *
+ * @returns 0 on success, else error
+ */
+
+int rm(const char *path) {
```
Comment (wkz) [inline src/confd/src/infix-services.c:461]: This seems like an odd place for such a generic function; is it a candidate for `libite`?

Also, it is quite intricate. Any reason we can't take the approach suggested on [SO](https://stackoverflow.com/a/5467788/715363)?

File: src/confd/yang/infix-services.yang [hunk 12]
```diff
@@ -104,6 +124,48 @@ module infix-services {
       }
     }
   }
+  container ssh {
+    description "Configuration for the SSH daemon";
+    leaf enabled {
+      description "Disable or enable SSH daemon";
```
Comment (wkz) [inline src/confd/yang/infix-services.yang:130]: Swapping disable/enable here would be more in line with the rest of our service descriptions.

**Review by wkz — CHANGES_REQUESTED** [review]
Great job! :muscle:

File: src/confd/src/infix-services.c (hunk 11 above)
Comment (troglobit) [reply src/confd/src/infix-services.c:461]: It was an omission from my side, I have a similar function in Finit that I haven't refactored and moved to libite yet.  Maybe we can use `libsrx/src/{helpers,common}.[ch]` for generic helpers that can be migrated?

File: doc/ChangeLog.md [hunk 13]
```diff
@@ -13,6 +13,8 @@ All notable changes to the project are documented in this file.
    as "reflector" and filtering of reflected services.  Issue #678
  - Review of default `sysctl` settings, issue #829
  - Upgrade Linux kernel to 6.12.3 (LTS)
+ - SSH Server is now configurable
+ - SSH Server and NETCONF Server uses the same SSH hostkey in factory-config.
```
Comment (troglobit) [inline doc/ChangeLog.md:17]: Should be on a single bullet.  The first line should add ", issue #441"

**Review by troglobit — CHANGES_REQUESTED** [review]
Nice work, only minor stuff to comment on. 

I guess the [implement](https://github.com/kernelkit/infix/pull/838/commits/5f0548e4952a8b61acf1cebf98f364010259cad0) commit should be squashed with the last commit that adds `rmrf()`, or split and squashed where one part is squashed with [Add support for configurable SSH server](https://github.com/kernelkit/infix/pull/838/commits/c224ec2063b747c1c3e504854e85925dbeb6df45)?

File: doc/ChangeLog.md [hunk 14]
```diff
@@ -13,6 +13,8 @@ All notable changes to the project are documented in this file.
    as "reflector" and filtering of reflected services.  Issue #678
  - Review of default `sysctl` settings, issue #829
  - Upgrade Linux kernel to 6.12.3 (LTS)
+ - SSH Server is now configurable
```
Comment (troglobit) [inline doc/ChangeLog.md:16]: Still missing reference to issue #441.

File: package/skeleton-init-finit/skeleton/etc/finit.d/available/sshd.conf [hunk 15]
```diff
@@ -1,4 +1,2 @@
-task    <pid/syslogd> \
-	[S]       /usr/bin/ssh-hostkeys                        -- Verifying SSH host keys
-service <task/ssh-hostkeys/success> env:-/etc/default/sshd \
+service <> env:-/etc/default/sshd \
```
Comment (troglobit) [inline package/skeleton-init-finit/skeleton/etc/finit.d/available/sshd.conf:1]: Please move the `<pid/syslogd>` condition to the main sshd servce now that the ssh-hostkeys start script is no longer used.

File: src/confd/src/ietf-keystore.c [hunk 16]
```diff
@@ -65,74 +54,98 @@ static char *filerd(const char *fn, size_t len)
...
+				continue;
+			private_key_format = srx_get_str(session, "%s/private-key-format", entry->xpath);
+			if (!private_key_format) {
+				free(public_key_format);
+				continue;
+			}
+
+			if (!strcmp(private_key_format, "ietf-crypto-types:rsa-private-key-format") &&
+			    !strcmp(public_key_format, "ietf-crypto-types:ssh-public-key-format"))
+			{
```
Comment (troglobit) [inline src/confd/src/ietf-keystore.c:93]: Minor coding style, malplaced brace.

File: src/confd/src/infix-services.c [hunk 17]
```diff
@@ -4,19 +4,26 @@
...
 #include <pwd.h>
 #include <stdarg.h>
+#include <sys/param.h>
 #include <sys/utsname.h>
 #include <sys/sysinfo.h>
 #include <sys/types.h>
 
 #include <srx/common.h>
+#include <srx/helpers.h>
+#include <dirent.h>
```
Comment (troglobit) [inline src/confd/src/infix-services.c:14]: Do we still need the sys/param.h and dirent.h includes with the new `rmrf()`?

File: src/confd/yang/infix-services.yang (hunk 5 above)
Comment (troglobit) [reply src/confd/yang/infix-services.yang:17]: [RFC 9640](https://www.rfc-editor.org/rfc/rfc9640.html) or pre-RFC draft?

**Review by troglobit — CHANGES_REQUESTED** [review]
Awesome, getting there!  A few comments, and some still outstanding, added/updated.

Also, as per our discord discussion.  It would be great to at least get an example of how-to-add-ssh to my custom factory-config, here https://github.com/kernelkit/infix/blob/main/doc/branding.md#example-snippets

File: doc/ChangeLog.md (hunk 14 above)
Comment (mattiaswal) [reply doc/ChangeLog.md:16]: had forget to push.....

**Review by wkz — APPROVED** [review]
:star2: :fire: :star2:

**Review by troglobit — APPROVED** [review]
Nice and clean, well done! :shipit:

**Pull request merged**